packages/vite/src/node/fsUtils.ts

```diff
@@ -0,0 +1,436 @@
+export function getFsUtils(config: ResolvedConfig): FsUtils {
+  let fsUtils = cachedFsUtilsMap.get(config)
+  if (!fsUtils) {
+    if (config.command !== 'serve' || !config.server.fs.cachedChecks) {
+      // cached fsUtils is only used in the dev server for now, and only when the watcher isn't configured
+      // we can support custom ignored patterns later
+      fsUtils = commonFsUtils
+    } /* TODO: Enabling for testing, we need to review if this guard is needed
+    else if (config.server.watch === null || config.server.watch?.ignored) {
+      config.logger.warn(
+        colors.yellow(
+          `${colors.bold(
+            `(!)`,
+          )} server.fs.cachedChecks isn't supported if server.watch is null or a custom server.watch.ignored is configured\n`,
+        ),
+      )
+      fsUtils = commonFsUtils
+    } */ else if (
+      !config.resolve.preserveSymlinks &&
+      config.root !== getRealPath(config.root)
+    ) {
+      config.logger.warn(
+        colors.yellow(
+          `${colors.bold(
+            `(!)`,
+          )} server.fs.cachedChecks isn't supported resolve.preserveSymlinks is false and root is symlinked\n`,
+        ),
+      )
+      fsUtils = commonFsUtils
+    } else {
+      fsUtils = createCachedFsUtils(config)
+    }
+    cachedFsUtilsMap.set(config, fsUtils)
+  }
+  return fsUtils
+}
+
+type DirentsMap = Map<string, DirentCache>
+
+type DirentCacheType =
+  | 'directory'
+  | 'file'
+  | 'symlink'
+  | 'error'
+  | 'directory_maybe_symlink'
+  | 'file_maybe_symlink'
+
+interface DirentCache {
+  dirents?: DirentsMap
+  type: DirentCacheType
+}
+
+function readDirCacheSync(file: string): undefined | DirentsMap {
+  let dirents: fs.Dirent[]
+  try {
+    dirents = fs.readdirSync(file, { withFileTypes: true })
+  } catch {
+    return
+  }
+  return direntsToDirentMap(dirents)
+}
+
+function direntsToDirentMap(fsDirents: fs.Dirent[]): DirentsMap {
+  const dirents = new Map<string, DirentCache>()
+  for (const dirent of fsDirents) {
+    // We ignore non directory, file, and symlink entries
+    const type = dirent.isDirectory()
+      ? 'directory'
+      : dirent.isSymbolicLink()
+        ? 'symlink'
+        : dirent.isFile()
+          ? 'file'
+          : undefined
+    if (type) {
+      dirents.set(dirent.name, { type })
+    }
+  }
+  return dirents
+}
+
+function ensureFileMaybeSymlinkIsResolved(
+  direntCache: DirentCache,
+  filePath: string,
+) {
+  if (direntCache.type !== 'file_maybe_symlink') return
+
+  const isSymlink = fs
+    .lstatSync(filePath, { throwIfNoEntry: false })
+    ?.isSymbolicLink()
+  direntCache.type =
+    isSymlink === undefined ? 'error' : isSymlink ? 'symlink' : 'file'
+}
+
+function pathUntilPart(root: string, parts: string[], i: number): string {
+  let p = root
+  for (let k = 0; k < i; k++) p += '/' + parts[k]
+  return p
+}
+
+export function createCachedFsUtils(config: ResolvedConfig): FsUtils {
+  const root = withoutTrailingSlash(config.root)
```

Comment:
Note: I added `withoutTrailingSlash` because the rest of the cache expects this. I think `config.root` is always without a trailing slash because we do a `path.resolve(config.root) : path.cwd()`. At least in UNIX, these two always return without a trailing slash. @sapphi-red could you confirm this is also the case in Windows? We could add `withoutTrailingSlash` in another PR if not when resolving the root, then there are some places where we could directly use `root + pathInRoot` instead of a `path.join()` call.

Reply:
I confirmed that `path.resolve(variable)` and `process.cwd()` both always return without a trailing slash on Windows 👍

Reply:
Thanks for testing! I think I will remove the new `withoutTrailingSlash` function then so we can take any testing to also check this more in general.